### PR TITLE
Rem 246

### PR DESCRIPTION
**Pull request by Alinadimmokdad**
chart of account account Ref field

### CODE REVIEW

**Comment by NoraBitarSM** [comment]
the field should be mandatory and should have a label

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 1]
```diff
@@ -47,37 +31,48 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
+      recordId: null,
+      accountRef: '',
+      name: '',
+      description: '',
+      groupId: '',
+      isCostElement: false,
+      sign: '',
+      groupName: '',
+      activeStatus: '',
+      activeStatusName: ''
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:45]: no need for groupName and activeStatusName in the formik instance

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 2]
```diff
@@ -47,37 +31,48 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
+    },
+    maxAccess: maxAccess,
     enableReinitialize: true,
     validateOnChange: true,
     validationSchema: yup.object({
       name: yup.string().required(' '),
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:51]: status is required also, try to save without status, you will get an error

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 3]
```diff
@@ -47,37 +31,48 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
-      description: yup.string().required(' ')
+      description: yup.string().required(' '),
+      accountRef: yup
+        .string()
+        .required()
+        .test(
+          'is-mask-filled',
+          'Account Reference is incomplete',
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:58]: no need for this text, it won't be shown anyway

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 4]
```diff
@@ -88,57 +83,48 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
-      } catch (exception) {
-        setErrorMessage(error)
-      }
-      setIsLoading(false)
-    })()
-  }, [])
-
-  const [segments, setSegments] = useState([])
+    if (recordId) {
+      setIsLoading(true)
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:87]: no need for this

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 5]
```diff
@@ -25,19 +21,7 @@ import { SystemRepository } from 'src/repositories/SystemRepository'
 export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
   const [isLoading, setIsLoading] = useState(false)
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:22]: no need

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 6]
```diff
@@ -88,57 +83,48 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
+          setSegments(
+            res.record.accountRef.split('-').map((seg, index) => ({
+              key: `GLACSeg${index}`,
+              value: seg.length
+            }))
+          )
+          setIsLoading(false)
+        })
+        .catch(error => {
+          toast.error('Failed to fetch record')
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:103]: ?

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 7]
```diff
@@ -170,8 +156,10 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
             segments={segments}
             name='accountRef'
             setFieldValue={formik.setFieldValue}
-            values={formik.values.accountRef.split('-')}
-            error={formik.touched.name && Boolean(formik.errors.name)}
+            values={formik.values.accountRef}
+            label='Account Reference'
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:160]: set the label from labels variable not hardcoded

File: src/pages/gl-chart-of-accounts/index.js [hunk 8]
```diff
@@ -11,27 +11,23 @@ import GridToolbar from 'src/components/Shared/GridToolbar'
...
 // ** API
 import { RequestsContext } from 'src/providers/RequestsContext'
-
 import { GeneralLedgerRepository } from 'src/repositories/GeneralLedgerRepository'
 
-// ** Windows
-import ChartOfAccountsWindow from './windows/ChartOfAccountsWindow'
-
 // ** Helpers
 import ErrorWindow from 'src/components/Shared/ErrorWindow'
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/index.js:17]: do u need this ?

File: src/pages/gl-chart-of-accounts/index.js [hunk 9]
```diff
@@ -11,27 +11,23 @@ import GridToolbar from 'src/components/Shared/GridToolbar'
...
 const ChartOfAccounts = () => {
   const { getRequest, postRequest } = useContext(RequestsContext)
- 
-  const [selectedRecordId, setSelectedRecordId] = useState(null)
+
+  const { stack } = useWindow()
 
   //states
-  const [windowOpen, setWindowOpen] = useState(false)
   const [errorMessage, setErrorMessage] = useState(null)
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/index.js:31]: ?

File: src/pages/gl-chart-of-accounts/index.js [hunk 10]
```diff
@@ -116,40 +102,54 @@ const ChartOfAccounts = () => {
...
-            setWindowOpen(false)
-            setSelectedRecordId(null)
-          }}
-          labels={_labels}
-          maxAccess={access}
-          recordId={selectedRecordId}
 
-        />
-      )}
       <ErrorWindow open={errorMessage} onClose={() => setErrorMessage(null)} message={errorMessage} />
```
Comment (NoraBitarSM) [inline src/pages/gl-chart-of-accounts/index.js:153]: ?

**Review by NoraBitarSM — CHANGES_REQUESTED** [review]

File: src/pages/gl-settings/index.js [hunk 11]
```diff
@@ -271,6 +271,7 @@ const GLSettings = () => {
               <Grid key={name} item xs={12} sx={{ marginTop: '7px' }}>
                 <CustomTextField
                   name={name}
+                  label={`Segment Name ${idx + 1}`}
```
Comment (hadichahine) [inline src/pages/gl-settings/index.js:274]: let's move this to labels

File: src/pages/gl-settings/index.js [hunk 12]
```diff
@@ -8,9 +8,8 @@ import CustomTextField from 'src/components/Inputs/CustomTextField'
 import { RequestsContext } from 'src/providers/RequestsContext'
 import { SystemRepository } from 'src/repositories/SystemRepository'
 import { ResourceIds } from 'src/resources/ResourceIds'
-import { useResourceQuery } from 'src/hooks/resource'
+import { useInvalidate, useResourceQuery } from 'src/hooks/resource'
```
Comment (hadichahine) [inline src/pages/gl-settings/index.js:11]: Do w need this?

File: src/pages/gl-chart-of-accounts/index.js [hunk 13]
```diff
@@ -102,6 +95,28 @@ const ChartOfAccounts = () => {
     } catch (err) {}
   }
 
+  const edit = obj => {
+    openForm(obj.recordId)
+  }
+
+  const add = () => {
+    openForm('')
```
Comment (hadichahine) [inline src/pages/gl-chart-of-accounts/index.js:103]: Remove empty strings.

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 14]
```diff
@@ -80,57 +74,43 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
+            }))
+          )
+        })
+        .catch(error => {})
+    } else {
+      getDataResult()
+    }
+  }, [recordId])
 
   const getDataResult = () => {
```
Comment (hadichahine) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:97]: Rename this to get latest segment data.

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 15]
```diff
@@ -240,58 +225,50 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
-    const newValues = [...values]
-    newValues[index] = event.target.value.slice(0, segments[index].value)
-    setFieldValue('segments', newValues)
-
-    const finalInput = newValues.join('-')
-    setFieldValue(name, finalInput)
-
-    if (event.target.value.length >= segments[index].value && index < segments.length - 1) {
-      inputRefs[index + 1].current.focus()
+const SegmentedInput = ({ segments, name, setFieldValue, values, label, error, helperText, required, readOnly }) => {
```
Comment (hadichahine) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:228]: Place next of Chartofaccountsform. Not nested.

File: src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js [hunk 16]
```diff
@@ -240,58 +225,50 @@ export default function ChartOfAccountsForm({ labels, maxAccess, recordId }) {
...
+        inputComponent={InputMask}
+        inputProps={{
+          mask: mask,
+          alwaysShowMask: true,
+          guide: false,
+          readOnly: readOnly
+        }}
+        required={required}
+      />
+      {error && <FormHelperText>{helperText}</FormHelperText>}
```
Comment (hadichahine) [inline src/pages/gl-chart-of-accounts/forms/ChartOfAccountsForm.js:271]: Check if need this.

**Review by hadichahine — CHANGES_REQUESTED** [review]

**Review by hadichahine — APPROVED** [review]

**Pull request merged**